core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -141,33 +138,49 @@ private List<Result> scanInternal(Scan originalScan) throws CrudException {
 
     Optional<Map<Snapshot.Key, TransactionResult>> resultsInSnapshot = snapshot.getResults(scan);
     if (resultsInSnapshot.isPresent()) {
-      return createScanResults(scan, originalProjections, resultsInSnapshot.get());
+      return createFilteredScanResults(scan, originalProjections, resultsInSnapshot.get());
     }
 
+    // Results for users
     Map<Snapshot.Key, TransactionResult> results = new LinkedHashMap<>();
 
+    // Results for the scan set
+    Map<Snapshot.Key, TransactionResult> actualResults = new LinkedHashMap<>();
+
     Scanner scanner = null;
     try {
-      scanner = scanFromStorage(scan);
-      for (Result r : scanner) {
-        TransactionResult result = new TransactionResult(r);
-        if (!result.isCommitted()) {
-          throw new UncommittedRecordException(
-              scan,
-              result,
-              CoreError.CONSENSUS_COMMIT_READ_UNCOMMITTED_RECORD.buildMessage(),
-              snapshot.getId());
-        }
+      @Nullable Integer limit = null;
+      if (scan.getLimit() == 0) {
+        scanner = scanFromStorage(scan);
+      } else {
+        limit = scan.getLimit();
+
+        // Get a scanner without the limit if the scan has a limit
+        scanner = scanFromStorage(Scan.newBuilder(scan).limit(0).build());
+      }
 
+      for (Result r : scanner) {
         Snapshot.Key key = new Snapshot.Key(scan, r);
+        TransactionResult result = new TransactionResult(r);
+        actualResults.put(key, result);
+        processScanResult(key, scan, result).ifPresent(res -> results.put(key, res));
 
-        // We always update the read set to create before image by using the latest record (result)
-        // because another conflicting transaction might have updated the record after this
-        // transaction read it first.
-        snapshot.putIntoReadSet(key, Optional.of(result));
-
-        snapshot.getResult(key).ifPresent(value -> results.put(key, value));
+        // Get results up to the limit if the scan has a limit
+        if (limit != null && results.size() >= limit) {
+          break;
+        }
+      }
+    } catch (RuntimeException e) {
+      Exception exception;
+      if (e.getCause() instanceof ExecutionException) {
+        exception = (ExecutionException) e.getCause();
+      } else {
+        exception = e;
       }
+      throw new CrudException(
+          CoreError.CONSENSUS_COMMIT_SCANNING_RECORDS_FROM_STORAGE_FAILED.buildMessage(),
+          exception,
+          snapshot.getId());
```

Comment:
Added exception handling to account for cases where the scanner iterator may throw a `RuntimeException`.